FB8-229: MTS using table and row level dependencies (8.0.17)

inikep opened the pull request:
This PR consist of:
1. 34 commits ported from 5.6 at https://github.com/facebook/mysql-5.6/pull/1066/commits/70d144faa1324316053ceff21e6c6b442dd6566e
2. Improvements based on Abhinav's 8.0.11 patch at https://github.com/facebook/mysql-5.6/pull/1066/commits/392a9b66a3ec6d67d9b466dc3d3580b5de3a6f74, https://github.com/facebook/mysql-5.6/pull/1066/commits/e9f44a8332f8efb0ddb219c700e6198ed21bd4fd and https://github.com/facebook/mysql-5.6/pull/1066/commits/73d6ad305abe4233140af84418600b29cc0ae889
3. Upstream MTR parallel replication tests moved to `rpl_mts_gtid`, `rpl_mts_nogtid`, and `rpl_mts` (dependency replication).
4. Some upstream MTR parallel replication tests were fixed: https://bugs.mysql.com/bug.php?id=97694 https://bugs.mysql.com/bug.php?id=97708 https://bugs.mysql.com/bug.php?id=97713 https://bugs.mysql.com/bug.php?id=97715
5. Added `rpl_mts_dependency_validate` to check correctness of dependency-based replication

inikep committed:
Move all MTS tests from "rpl_gtid" to a new suite "rpl_mts_gtid"

inikep committed:
Move all MTS tests from "rpl" and "rpl_nogtid" to a new suite "rpl_mts_nogtid"

inikep committed:
Move some MTS tests from "rpl_mts_gtid" to "rpl_mts"

inikep committed:
Move some MTS tests from "rpl_mts_nogtid" to "rpl_mts"

inikep committed:
Squashed implementation of MTS using table and row level dependencies

commit 9451944890fb871c3306fd96cbf52809d28dc23d
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Fri Nov 15 12:54:13 2019 -0800

    Clearing coordinator's mem root on every end event

    Summary:
    Vanilla MTS clears the rli's mem root (which is just preallocated block
    of memory in THD) after encountering an end event in
    Log_event::get_slave_worker(). We need to do the same in dependency replication.

    The coordinator's mem root is usually not used since it's used only by threads
    that execute transactions. In dependency replication however, we create a
    conversion table when the schema between the master and slave is different to
    parse out keys correctly. This conversion table is allocated in the mem root.
    Note that not clearing mem root is not really a memory leak since it will be
    cleared when the coordinator is stopped but it still causes rss memory to
    increase.

    Reviewed By: anirbanr-fb

    Differential Revision: D18479423

    fbshipit-source-id: f8289664ca9

commit f0c1a59ff2a96975c45fd32e1c7e478a7397e98a
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Wed Nov 6 15:23:13 2019 -0800

    wait_for_dep_workers_to_finish should return error

    Summary:
    If the thread waiting for workers to finish is killed we should return
    an error and handle it appropriately.

    Reviewed By: anirbanr-fb

    Differential Revision: D18250470

    fbshipit-source-id: 29abade880c

commit 85c6ed5ee398672aacf94a669f53c77bd9577b6d
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Fri Aug 2 08:55:47 2019 -0700

    Fix partial trx condition when waiting for workers to finish

    Summary:
    After per DB commit ordering in D13086060, we buffer transactions in the SQL
    thread until we find an event that contains DB information. While syncing
    transactions (like DDL which need to be executed in isolation) we need to use
    the `trx_queued` flag to figure out if we need to factor in a partial
    transaction (i.e. a transaction  which is not fully parsed by SQL thread yet),
    otherwise race conditions can occur because SQL thread will not wait for all
    transactions to complete before queuing the sync transaction. In the case of DDL,
    the race condition can cause the DDL to run before a conflicting trx, this will cause
    a deadlock because of commit ordering or MDL locks.

    Reviewed By: anirbanr-fb

    Differential Revision: D16590942

    fbshipit-source-id: 32ef00f1335

commit d2329cc28b64dab090a2358662faf87a66d9520b
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Tue Apr 9 10:43:56 2019 -0700

    Handle column type conversion during key parsing

    Summary:
    We were not handling column type conversions while parsing keys to
    calculate dependencies. Copied code from slave worker which handles this while
    applying row changes to make it work.

    Reviewed By: anirbanr-fb

    Differential Revision: D14523433

    fbshipit-source-id: 397dd68c09a

commit b293358bc1c4620e2cbccda42f87ddcdc40cfa25
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Fri Jul 28 13:00:43 2017 -0700

    Fix flaky rpl_slave_tx_isolation

    Summary: Waiting for slave threads to reach a state before querying

    Reviewed By: tianx

    Differential Revision: D5518737

    fbshipit-source-id: 2617d25

commit 7d8d43781043929f5897fce1e0488936afa0a6ca
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Wed Oct 31 15:47:50 2018 -0700

    Wake workers up before waiting for them to stop

    Summary:
    While stopping slave workers we should first wake them up and then wait
    for them to exit. We used to do this in lock step for each worker which can
    cause us to wait on a single worker for ever when that worker depends on a
    different worker which has not been woken up yet.

    Reviewed By: anirbanr-fb

    Differential Revision: D12851975

    fbshipit-source-id: 3036bdc4c53

commit 76abc99fd60309c0e42eddfa187d4310a9193da2
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Wed Sep 19 13:19:23 2018 -0700

    Compare non-binary keys correctly to find dependencies

    Summary:
    For unique indices that contain columns with non-binary collations like
    latin1_swedish_ci we need cannot simply use memcmp between column values to find
    dependencies. For example, for any case insensitive collation 'a' and 'A' should
    conflict but memcmp will not detect this. To fix this we use
    `Field::make_sort_key()` method which returns a massaged column value which can
    later be compared using memcmp. In most cases `Field::make_sort_key()` uses
    `MY_COLLATION_HANDLER::strnxfrm()` to calculate the value. Originally this
    method is used to sort keys.

    Reviewed By: anirbanr-fb

    Differential Revision: D9946669

    fbshipit-source-id: 6dc86711f89

commit d2fe78b1d4a3c92e60dd0c8db3e092e9467f7d33
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Wed Aug 22 11:55:55 2018 -0700

    Fix race in slave worker running status during stop slave

    Summary:
    There was a race between `handle_slave_worker()` setting running_status
    to NOT_RUNNING and `slave_stop_worker()` checking the status. Cleaned up the
    code to obtain `jobs_lock` and wait for `jobs_cond` while waiting for workers to
    quit.

    Squash with D9073572

    Reviewed By: Tema

    Differential Revision: D9444080

    fbshipit-source-id: ae6dacfa134

commit 7eab374b73a7716bd9b110e4735e234c3dd3488b
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Wed Apr 18 11:09:53 2018 -0700

    Fix rpl_parallel_switch_sequential

    Summary:
    Dependency replication sometimes kills the query on the workers to stop
    the slave. Adding a supression to ignore killed query errors.

    Reviewed By: hermanlee

    Differential Revision: D7671553

    fbshipit-source-id: 7bb938e

commit 9599ca522ecfaf0fef18362294b388747414e90f
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Mon Dec 25 16:10:29 2017 -0800

    Cleaning up PSI during table cache cleanup in dep repl

    Summary:
    #title

    Squash with: D6138167

    Reviewed By: lth, yoshinorim

    Differential Revision: D6635282

    fbshipit-source-id: a8a9412

commit b70c79e1b014feb3bcec787e93662ac5f881a918
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Thu Aug 3 13:07:21 2017 -0700

    Fixing rpl.rpl_parallel_switch_sequential and rpl_mts.rpl_find_binlog_by_gtid

    Summary:
    rpl.rpl_parallel_switch_sequential had a typo in a variable that sets a
    timeout. rpl_mts.rpl_find_binlog_by_gtid was running a large workload
    unnecessarily. Both were timing out in valgrind.

    Reviewed By: yashtc

    Differential Revision: D5555630

    fbshipit-source-id: f182987

commit e48c599dfe4fbf149bc21fd54177b56c35ab9a80
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Wed Aug 2 15:12:44 2017 -0700

    Fix timeout issues in rpl_mts_dependency_order_commits

    Summary: Increased timeout, will check if this works

    Reviewed By: teng-li

    Differential Revision: D5547923

    fbshipit-source-id: 424d311

commit e7963d7c4ba327d4e757e0c42ce0c4ca69a41622
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Thu Jun 13 08:23:33 2019 -0700

    Cache global variables related to dep repl in rli for thread safety

    Summary:
    We were using global variables directly, this might cause the slave to
    crash if `STOP SLAVE` was not executed before setting the variables.

    Reviewed By: anirbanr-fb

    Differential Revision: D15771019

    fbshipit-source-id: 7496e79acf8

commit 51a625078b58ef2d17ac1ce78af40bb30a9729b5
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Thu Apr 4 09:48:38 2019 -0700

    Sync transaction when number of keys exceed a threshold

    Summary:
    When a transaction modifies too many rows the map which is used to hold
    keys (to find dependencies) can become too large. This results in slow lookups
    and memory pressure. This diff introduces a new variable mts_dependency_max_keys
    (default: 100000) that defines a threshold after which the transaction is
    executed in isolation (i.e. we wait for all transactions before it to complete
    and then execute it isolation). Since the transaction is executed in isolation,
    its key information is not stored

    Also changed the hash function for keys to well known murmur hash which has a
    better distribution.

    Reviewed By: anirbanr-fb

    Differential Revision: D14235731

    fbshipit-source-id: fabe4ceb2f2

commit 41a26b8461dca9dc978f0c7d41a853b753b44708
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Thu Jan 3 09:26:19 2019 -0800

    Per database commit ordering

    Summary:
    Ordering commits across databases is too strict and not required for
    most use cases. This implementation switches to commit ordering within databases
    to improve slave throughput.

    Reviewed By: anirbanr-fb

    Differential Revision: D13086060

    fbshipit-source-id: 2134da98596

commit 65dc20b6962f93c126373042d914ee65adf43313
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Fri Aug 17 19:50:19 2018 -0700

    Handle multi-table update statements

    Summary:
    We used to maintain only the last seen table map event to compute
    dependencies. In multi-table updates like:
    UPDATE t1, t2 SET a = a * 10, b = b * 10 WHERE t1.a = t2.b;
    both tables may be updated which leads to two consecutive table map events
    followed by all rows events of both table.

    To fix this, we now maintain a map of table id -> table map event and then
    extract the correct table map event using the table id info in rows event to
    calculate dependencies.

    Also removed the dependency between global vars
    slave_use_idempotent_for_recovery and mts_dependency_replication. Now we only
    print an error in the log when idempotent recovery is not enabled while using
    dependency replication.

    Reviewed By: anirbanr-fb

    Differential Revision: D9105753

    fbshipit-source-id: 6f5d6382f40

commit 32462fb5c0d2fbbbad219ef15066af9dc3e31db1
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Fri Aug 17 19:37:47 2018 -0700

    Worker threads should respond to kill command in any state

    Summary:
    There were some states where worker threads were not responsive to kill
    command. Fixed those cases. Also added a stop slave stress test.

    Reviewed By: anirbanr-fb

    Differential Revision: D9073572

    fbshipit-source-id: 92da9fc0c8f

commit 5a7c298f34da9b828fd06de03907e3a1e21a00e5
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Fri Jun 29 11:37:19 2018 -0700

    Improve signalling in dependency replication

    Summary:
    Signal only when a worker is waiting to save CPU cycles. Also adding
    likely/unlikey in some parts of the code that is called very frequently.

    Reviewed By: anirbanr-fb

    Differential Revision: D8408430

    fbshipit-source-id: 52adc8e

commit b67891e57f1fa9bfddae6713fb4043933f273e05
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Mon Apr 30 14:22:47 2018 -0700

    Cleanup key parsing and handle unique key conflicts

    Summary:
    Unique keys must be considered in conflict detection between transactions. A
    simple example is:

    1. CREATE TABLE t1(a INT PRIMARY KEY, b INT UNIQUE KEY);
    2. INSERT INTO t1 VALUES(1, 1);
    3. UPDATE T1 SET b = 2 WHERE a = 1;
    4. INSERT INTO t1 VALUES(2, 1);

    If we only depend on PK schedule 2, 4, 3 could run because 3 would be
    independent, but this schedule will cause a duplicate key error on UK b.

    Also removing any code that switches from stmt to table dependencies in
    exceptional circumstances. Now we're just syncing the entire group/trx (i.e.
    executing is isolation) which makes the code much cleaner.

    Depends on D7593232

    Reviewed By: anirbanr-fb

    Differential Revision: D7596392

    fbshipit-source-id: cebbfd5

commit 4a6e25892f01b66c4b5d7b404fce56eadf331b60
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Tue Mar 20 06:01:58 2018 -0700

    Reduce contention in dependency replication

    Summary:
    Replace the global DAG by a lightweight producer-consumer queue and track
    dependencies within individual events.

    Reviewed By: anirbanr-fb

    Differential Revision: D7108777

    fbshipit-source-id: a94579c

commit b0857df5665f69f9d020dd9125f8a1d824e53423
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Tue Jul 25 21:28:45 2017 -0700

    New global variable `slave_tx_isolation` to control trx isolation level for slave replication threads.

    Summary:
    Dependency replication takes care of write-write conflicts in
    trxs using explicit dependencies. Isolation levels like repeatable-read can
    cause snapshot errors when two transactions whose write-sets conflict but whose
    execution on the master don't overlap (in time) are scheduled in parallel on the
    slave. These snapshot errors cause the slave to rollback and retry transactions
    which can lower the throughput considerably.

    This feature enables slave replication threads to run on a isolation level
    specified by variable `slave_tx_isolation`. The slave need to be restarted
    before the effects can be seen.

    Reviewed By: tianx

    Differential Revision: D5482252

    fbshipit-source-id: 76d77a0

commit 245589fdf55f589cf110f68ef5f31505d50eecf1
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Tue Jan 9 15:44:13 2018 -0800

    Fix stmt level dependency replication

    Summary:
    Fixed a bunch of things:

    1. Used both before and after images of update events to find both RW and WW
    conflicts
    2. Added a flag to indicate a worker failure and stopped other workers from
    executing events if the flag is on immediatly
    3. Used references in for-loops for better performance
    4. Used READ-COMMITTED in dep-stmt and dep-tbl to reduce trx retries on slaves

    Squash with: D6138167

    Reviewed By: hermanlee

    Differential Revision: D6681471

    fbshipit-source-id: ffa0059

commit 8633b0bbba55c72df6fdf1b243609cfd2bcc8c9e
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Sat Dec 23 16:47:23 2017 -0800

    Fixing test cases affected by type change of mts_dependency_replication

    Summary:
    mts_dependency_replication was changed from a boolean variable to an
    enum. Modifying and re-recording test cases accordingly.

    Squash with: D6138167

    Reviewed By: anirbanr-fb

    Differential Revision: D6633752

    fbshipit-source-id: ddeefa4

commit a5361bfa25679428dcf595612e57b313aeb1ecb2
Author: Jose Faleiro <jfaleiro@fb.com>
Date:   Fri Dec 22 14:34:26 2017 -0800

    Tracking dependencies between row events using primary keys

    Summary:
    With table level dependencies two row events were dependent on each other if
    they modified the same table. This caused unnecessary dependencies between row
    events which touch completely different rows in the same table. This change
    parses out primary keys from rows events to find dependencies i.e. two row
    events will conflict only when the sets of primary keys they modify intersect,
    otherwise they will be executed completely independently.

    Reviewed By: abhinav04sharma

    Differential Revision: D6138167

    fbshipit-source-id: 8cfc9b9

commit 3cfef9bf5b3c4e55ad606a7436bbb2f30e51b429
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Sun Dec 10 20:48:40 2017 -0800

    Fixed bug in identifying table map event in the SQL thread for dep replication

    Summary:
    Dependency replication extracts the full table name from table
    map event. Usually, table map event is always written right before every rows
    event in the binlog, but some special statements like REPLACE INTO can
    generate both update and write rows events which are written after a single
    table map event. This change stores the last table map event encountered by the
    SQL thread a separate variable which is used instead of relying on
    Relay_log_info::prev_event to get the associated table map event for
    every rows event.

    Squash with: D6389969

    Reviewed By: anirbanr-fb

    Differential Revision: D6531543

    fbshipit-source-id: 057ccfc

commit 0b9ccc83b5a3058d4ad4cb1327d1e27c2f93ff98
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Tue Nov 21 21:43:58 2017 -0800

    Using table names instead of table IDs to identify dependencies

    Summary:
    Table IDs are not guaranteed to be unique. They are assigned to tables
    using an in-memory counter that is incremented everytime a new table is opened.
    So, under certain circumstances like master reboot the same table can have
    multiple table IDs.

    This change uses database name + table name to identify dependencies instead of
    table IDs.

    Reviewed By: anirbanr-fb

    Differential Revision: D6389969

    fbshipit-source-id: f90c925

commit fe4e9075870f1ff64cd8375c10a1a0d78f85311b
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Thu Oct 5 13:37:30 2017 -0700

    Fixed setting condition variables related to DAG occupancy

    Summary:
    Some condition variables were set after unlocking the DAG. Any changes
    to condition vars related to the DAG should be done after securing a lock on the
    DAG so that the values of these vars correctly reflect the state of the DAG

    Reviewed By: anirbanr-fb

    Differential Revision: D5982946

    fbshipit-source-id: 256a0ee

commit 69046525e2b3f4e00060d14bf7a85dfe812e638c
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Fri Jul 28 15:49:10 2017 -0700

    Perf improvements in dependency replication

    Summary:
    - Taking a lock on the DAG once to extract the entire trx/group instead of
      taking a lock to extract each event
    - Sending signal for only a begin event that is ready between the DAG
      mutators and the consumers, previously the signal was sent for every event
      even though the consumer only waits for a ready begin event
    - Fixed a memory leak of log events by using `is_appended_to_queue`
      instead of `is_assigned`

    Reviewed By: tianx

    Differential Revision: D5484833

    fbshipit-source-id: ed26ae6

commit 6a17633a5ceac106a948456701b5f4d766a3b578
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Tue Jun 27 17:30:12 2017 -0700

    Fixed valgrind errors related to `Log_event::contains_partition_info` and fixed a assert in `Dependency_slave_worker::remove_event()`

    Summary:
    Valgrind was complaining because `Relay_log_info::mts_end_group_sets_max_dbs` was used uninitialized and the assert in `remove_event()` was executed without taking a lock on the DAG.

    Squash with: D4382612

    Reviewed By: alxyang

    Differential Revision: D5335781

    fbshipit-source-id: da331d2

commit 62ff4bc5468242c786be4b77e9122828c5c0f402
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Thu Apr 27 21:26:17 2017 -0700

    Commit ordering for parallel replication

    Summary:
    Ported 5.7s commit order manager and made nesessary changes to
    integrate it with dependency tracking

    Depends on D4382612

    Reviewed By: anirbanr-fb

    Differential Revision: D4980434

    fbshipit-source-id: a281c98

commit e0d1ae2a30dbf3560e3c78b4cd00b493f4474b0c
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Fri Dec 2 12:28:09 2016 -0800

    Implementation of MTS using table level dependencies

    Summary:
    Using DAG of log events and pull based slave workers to execute events
    in parallel on the slave useing table level dependencies.

    Depends on D4302984

    Reviewed By: teng-li

    Differential Revision: D4382612

    fbshipit-source-id: a0ec32a

commit bdab751b002f162b8f98862aee2bc85174ea2d88
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Fri Dec 2 12:28:09 2016 -0800

    Implementation of slave worker that pulls events from DAG

    Summary:
    Slave worker implementation which pulls and executes log events
    from DAG in serial order. The worker pulls events in such a way that
    once it encounters a begin event of a group it pulls events to
    finish that group.

    Depends on D4302854

    Reviewed By: tianx

    Differential Revision: D4302984

    fbshipit-source-id: 10a0f11

commit c854f15e5957fdcd8df5404b029f491c1edb142c
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Thu Dec 8 11:35:01 2016 -0800

    Implementation of a DAG container

    Summary:
    Implementation of a simple DAG container which can be use to
    track dependencies between arbitrary objects

    Depends on D4980348

    Reviewed By: jkedgar

    Differential Revision: D4302854

    fbshipit-source-id: b87758e

commit d90323ff6c6a8eb97e875cab50da8d1af62b40e0
Author: Abhinav Sharma <abhinavsharma@fb.com>
Date:   Mon May 1 10:23:25 2017 -0700

    Moved all tests related to MTS to a new suite "rpl_mts"

    Summary: Moved all MTS test cases to a separate suite

    Reviewed By: teng-li

    Differential Revision: D4980348

    fbshipit-source-id: 56dfdef

inikep committed:
Introduce slave_parallel_type=DEPENDENCY

inikep committed:
Move dependency tracking related variables from Relay_log_info to Mts_submode_dependency

abhinav04sharma commented on sql/rpl_gtid_state.cc:
```diff
@@ -482,7 +483,17 @@ enum_return_status Gtid_state::generate_automatic_gtid(
+        automatic_gtid.sidno = global_sid_map->add_sid(db_uuid);
+      } else
+        automatic_gtid.sidno = get_server_sidno();
+    }
```
Is this change related to parallel replication?

abhinav04sharma commented on sql/rpl_mts_submode.cc:
```diff
@@ -164,7 +166,7 @@ int Mts_submode_database::wait_for_workers_to_finish(Relay_log_info *rli,
     // resources relocation
     mts_move_temp_tables_to_thd(thd, entry->temporary_tables);
-    entry->temporary_tables = nullptr;
+    entry->temporary_tables = NULL;
```
Is this change required?

abhinav04sharma commented on sql/rpl_mts_submode.cc:
```diff
@@ -958,3 +960,440 @@ Mts_submode_logical_clock::get_server_and_thread_id(TABLE *table) {
+    dep_queue.pop_front();
+  } else {
+    ret = nullptr;
+  }
```
nit: this should not be required.

abhinav04sharma commented on sql/rpl_reporting.cc:
```diff
@@ -86,8 +86,13 @@ int Slave_reporting_capability::has_temporary_error(THD *thd, uint error_arg,
   */
+  // Since slave_gtid_info table is not used if
+  // slave_use_idempotent_for_recovery is enabled, we don't consider COMMIT
+  // errors temporary if idempotent recovery is on.
```
Since we haven't ported `slave_gtid_info` into 8.0 this comment is a little misleading. Actually, I'm not sure if we really need this change...

inikep replied:
It was removed

abhinav04sharma commented on sql/log_event.h:
```diff
@@ -2843,6 +2905,21 @@ class Rows_log_event : public virtual binary_log::Rows_event, public Log_event {
+
+ protected:
+  bool parse_keys(Relay_log_info *rli, std::shared_ptr<Log_event_wrapper> &ev,
+                  TABLE *table, std::deque<Dependency_key> &keys);
```
Any reason for this method to be protected instead of private?

abhinav04sharma commented on sql/log_event_wrapper.cc:
```diff
@@ -0,0 +1,65 @@
+
+  mysql_mutex_lock(&mutex);
+  Slave_worker *worker = raw_ev_worker;
+  DBUG_ASSERT(worker);
```
Currently this assert is failing for `Rows_query_log_event` because `Rows_query_log_event::do_apply_event()` sets the worker to nullptr.

abhinav04sharma commented on sql/mysqld.cc:
```diff
@@ -1177,6 +1178,7 @@ const char *binlog_error_action_list[] = {"IGNORE_ERROR", "ABORT_SERVER",
 uint32 gtid_executed_compression_period = 0;
 bool opt_log_unsafe_statements;
 bool opt_log_global_var_changes;
+bool use_db_uuid = false;
```
Is this change related to this feature?

abhinav04sharma commented on sql/mysqld.cc:
```diff
@@ -11219,6 +11306,9 @@ PSI_stage_info stage_suspending= { 0, "Suspending", 0, PSI_DOCUMENT_ME};
 PSI_stage_info stage_hook_begin_trans= { 0, "Executing hook on transaction begin.", 0, PSI_DOCUMENT_ME};
+PSI_stage_info stage_slave_waiting_for_dependencies= { 0, "Waiting for dependencies to be satisfied", 0, PSI_DOCUMENT_ME};
+PSI_stage_info stage_slave_waiting_for_dependency_workers= { 0, "Waiting for dependency workers to finish", 0, PSI_DOCUMENT_ME};
+PSI_stage_info stage_slave_waiting_semi_sync_ack= { 0, "Waiting for an ACK from semi-sync ACKers", 0, PSI_DOCUMENT_ME};
```
Looks like this change also does not belong to this PR.

abhinav04sharma commented on sql/rpl_rli_pdb.cc:
```diff
@@ -2724,3 +2750,147 @@ const uint *Slave_worker::get_table_pk_field_indexes() {
+  }
+
+  error = worker->slave_worker_exec_event(ev);
+  ev_wrap->set_raw_ev_worker(ev);
```
Can we pass `worker` to this method instead? Or not maintain a separate `worker` member in `Log_event_wrapper` and just use the one in `Log_event`? FYI, there is an issue with the current code where worker will be null at this point for `Rows_query_log_event` (I've commented about this above).

inikep replied:
We can't use `Log_event` as it may be already freed in `Log_event_wrapper::next()`. I changed the function to accept `worker`.

abhinav04sharma commented on sql/rpl_slave.cc:
```diff
@@ -6930,6 +6964,33 @@ extern "C" void *handle_slave_sql(void *arg) {
+    commit_order_mngr =
+        new Commit_order_manager(rli->opt_slave_parallel_workers);
+
+  rli->set_commit_order_manager(commit_order_mngr);
```
This causes a mem-leak when both `slave_preserve_commit_order` and `mts_dependency_order_commits` are enabled (check line 6949 above). Maybe get rid of `mts_dependency_order_commits` completely?

abhinav04sharma commented on sql/rpl_slave_commit_order_manager.h:
```diff
@@ -1,27 +1,21 @@
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-   GNU General Public License, version 2.0, for more details.
```
Is this change required?

abhinav04sharma commented on sql/log_event.cc:
```diff
@@ -14521,3 +14560,421 @@ size_t my_strmov_quoted_identifier_helper(int q, char *buffer,
+    // for key parsing because upack_row takes cols as an argument instead of
+    // relying on read and write sets
+    bitmap_set_all(table->read_set);
+    bitmap_set_all(table->write_set);
```
Can these lines be moved outside the loop?

inikep replied:
It seems to work fine outside the loop. For example `rpl_mts_dependency_type_conv` fails without `bitmap_set_all` but works with `bitmap_set_all` moved outside the loop.

inikep committed:
Remove "mts_dependency_order_commits" in favor of "slave_preserve_commit_order"

inikep commented:
The changes in the updated patch:
1. All above comments were addressed
2. This patch was rebased to current `fb-mysql-8.0.17`
3. `rocksdb_rpl.rpl_mts_dependency_unique_key_conflicts` was enabled
4. Removed `mts_dependency_order_commits` in favor of `slave_preserve_commit_order` (warning: `slave_preserve_commit_order` is disabled by default)
5. Reverted https://github.com/facebook/mysql-5.6/commit/8fbbaff3f779 

I only tested (`--big-test --suite=rpl_mts`) the patch locally with clang-9 because `fb-mysql-8.0.17` has issues with gcc-9 compilation.

abhinav04sharma commented on sql/log_event.cc:
```diff
@@ -14535,3 +14574,424 @@ size_t my_strmov_quoted_identifier_helper(int q, char *buffer,
+
+  for (i = 0, curr_row = m_rows_buf; curr_row != m_rows_end;
+       curr_row = curr_row_end, i++) {
+    MY_BITMAP *cols = &m_cols;  // init with before image bitmap
```
Is this needed?

inikep replied:
Yes. I updated `Rows_log_event::parse_keys` for the case when the after image has a different set of columns than the before image. I'm not able to generate such a case but `Rows_log_event::force_write_to_binlog` supports it.
